### PR TITLE
Convert existing infrastructure into a standard c# project

### PR DESCRIPTION
**Pull request by TyOverby**
This PR converts all the existing infrastructure (from run-and-report downward) into a standard C# project built under the Roslyn solution.

Along with the project conversion, I also broke some files apart and reorganized some methods.

Here's a list of features that are supported so far:

* `.csx` file detection and test running
* CPC and Consumption integration for gathering traces and extracting metrics
* Upload to benchview
* Save traces to @basoundr's share

Everything else was removed from the filesystem (but still exists in git history). 
Those features will be ideally added back before merging, but could be added back later as well.

### CODE REVIEW

**Comment by TyOverby** [comment]
review please: @basoundr @rchande @KevinH-MS 

cc: @jaredpar

**Comment by TyOverby** [comment]
### In order to test this out yourself, run

```cmd
msbuild Roslyn.sln
.\Binaries\Debug\Roslyn.Test.Performance.Runner.exe
```

(I don't know if this will work if you don't have CPC installed yet, I haven't made a nuget for it)

File: src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj [hunk 1]
```diff
@@ -50,9 +50,9 @@
       <IncludeOutputGroupsInVSIXLocalOnly>DebugSymbolsProjectOutputGroup</IncludeOutputGroupsInVSIXLocalOnly>
```
Comment (basoundr) [inline src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj]: I guess you did intend to have these changes

File: src/Test/Perf/Runner/Benchview.cs [hunk 2]
```diff
@@ -1,22 +1,47 @@
...
-using System.Linq;
 using System.IO;
 using System.Xml;
 using static Roslyn.Test.Performance.Utilities.TestUtilities;
+using static Roslyn.Test.Performance.Runner.Tools;
 
-namespace Roslyn.Test.Performance.Utilities
+namespace Roslyn.Test.Performance.Runner
 {
-    internal class Tools
```
Comment (basoundr) [inline src/Test/Perf/Runner/Benchview.cs]: Doesn't the Roslyn analyzers complain that the Public element are missing XML doc comments?

File: src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj (hunk 1 above)
Comment (TyOverby) [inline src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj]: Correct. 

I also have no idea how this file got changed.  I never even opened it?!

Comment (TyOverby) [inline src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj]: That would explain why the tests were failing though.

File: src/Test/Perf/Runner/Benchview.cs (hunk 2 above)
Comment (TyOverby) [inline src/Test/Perf/Runner/Benchview.cs]: Yep, there's a bunch of complaining.  

I'm going to have a documentation commit after the structure is approved.

File: Roslyn.sln [hunk 3]
```diff
@@ -359,14 +359,20 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Moq.net", "src\Dependencies
...
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "EditorFeatures.Next", "src\EditorFeatures\Next\EditorFeatures.Next.csproj", "{366BBCDC-B05F-4677-9B5B-78BA816A1484}"
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "VisualStudioSetup.Next", "src\VisualStudio\Setup.Next\VisualStudioSetup.Next.csproj", "{143FE684-6E1C-41DF-9C60-84C7772DC49C}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Perf", "Perf", "{DD13507E-D5AF-4B61-B11A-D55D6F4A73A5}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Utilities", "src\Test\Perf\Utilities\Utilities.csproj", "{59AD474E-2A35-4E8A-A74D-E33479977FBF}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Runner", "src\Test\Perf\Runner\Runner.csproj", "{1AA6D2F0-2C40-4AF6-BB79-50AFDCC62720}"
```
Comment (jaredpar) [inline Roslyn.sln]: I understand these are scoped to the `Perf` folder but we should probably pick more unique names.  For instance `Perf.Utilities` and `Perf.Runner`.  In general project files need to be kept unique and these are pretty generic.

File: src/Test/Perf/Runner/Benchview.cs [hunk 4]
```diff
@@ -1,22 +1,47 @@
...
+    public class Benchview 
     {
+        public static readonly string CPCDirectoryPath = Environment.ExpandEnvironmentVariables(@"%SYSTEMDRIVE%\CPC");
+        private const string BenchviewPath = @"\\vcbench-srv4\benchview\uploads\vibench";
```
Comment (jaredpar) [inline src/Test/Perf/Runner/Benchview.cs]: should be prefixed with `s_`

File: src/Test/Perf/Runner/Benchview.cs [hunk 5]
```diff
@@ -1,22 +1,47 @@
...
+            else
+            {
+                Log("Conversion from Consumption to csv failed.");
+            }
+        }
+
         /// Takes a consumptionTempResults file and converts to csv file
         /// Each info contains the {ScenarioName, Metric Key, Metric value}
-        public static bool ConvertConsumptionToCsv(string source, string destination, string requiredMetricKey, ILogger logger)
+        static bool ConvertConsumptionToCsv(string source, string destination, string requiredMetricKey)
```
Comment (jaredpar) [inline src/Test/Perf/Runner/Benchview.cs]: Visibility modifier.

File: src/Test/Perf/Runner/Benchview.cs [hunk 6]
```diff
@@ -61,24 +86,24 @@ public static bool ConvertConsumptionToCsv(string source, string destination, st
...
+                Log(e.StackTrace);
                 return false;
             }
 
             return true;
         }
 
         /// Gets a csv file with metrics and converts them to ViBench supported JSON file
-        public static string GetViBenchJsonFromCsv(string compilerTimeCsvFilePath, string execTimeCsvFilePath, string fileSizeCsvFilePath, bool verbose, ILogger logger)
+        static string GetViBenchJsonFromCsv(string compilerTimeCsvFilePath, string execTimeCsvFilePath, string fileSizeCsvFilePath)
```
Comment (jaredpar) [inline src/Test/Perf/Runner/Benchview.cs]: Visibility modifier

File: src/Test/Perf/Runner/Program.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+using Roslyn.Test.Performance.Utilities;
+using static Roslyn.Test.Performance.Utilities.TestUtilities;
+using static Roslyn.Test.Performance.Runner.Tools;
+using static Roslyn.Test.Performance.Runner.Benchview;
+using static Roslyn.Test.Performance.Runner.TraceBackup;
+using Roslyn.Test.Performance.Runner;
+
+namespace Runner
+{
+    class Program
```
Comment (jaredpar) [inline src/Test/Perf/Runner/Program.cs]: Visibility modifier.  Looks like a lot are missing.  Please add them all

File: src/Test/Perf/Runner/Program.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
+﻿using System;
```
Comment (jaredpar) [inline src/Test/Perf/Runner/Program.cs]: Copyright header.

File: Roslyn.sln (hunk 3 above)
Comment (TyOverby) [inline Roslyn.sln]: Done.

File: src/Test/Perf/Runner/Benchview.cs (hunk 4 above)
Comment (TyOverby) [inline src/Test/Perf/Runner/Benchview.cs]: Done.

File: src/Test/Perf/Runner/Benchview.cs (hunk 5 above)
Comment (TyOverby) [inline src/Test/Perf/Runner/Benchview.cs]: Done.

File: src/Test/Perf/Runner/Benchview.cs (hunk 6 above)
Comment (TyOverby) [inline src/Test/Perf/Runner/Benchview.cs]: Done.

File: src/Test/Perf/Runner/Program.cs (hunk 7 above)
Comment (TyOverby) [inline src/Test/Perf/Runner/Program.cs]: I went through all the files adding visibility modifers.  I also made all of our utilities classes `static class`es.

File: src/Test/Perf/Runner/Program.cs (hunk 8 above)
Comment (TyOverby) [inline src/Test/Perf/Runner/Program.cs]: Done on all files.

**Comment by TyOverby** [comment]
@dotnet-bot test ci please

File: cibuild.cmd [hunk 9]
```diff
@@ -78,6 +79,12 @@ if defined TestDeterminism (
     exit /b 0
 )
 
+if defined TestPerfCorrectness (
+    msbuild %MSBuildAdditionalCommandLineArgs% Roslyn.sln /p:Configuration=%BuildConfiguration% || goto :BuildFailed
+    .\Binaries\%BuildConfiguration%\Roslyn.Test.Performance.Runner.exe
```
Comment (jaredpar) [inline cibuild.cmd]: Shouldn't htis have `|| goto :BuildFailed`?  As is it looks like this will successfully complete 100% of the time.

Comment (TyOverby) [inline cibuild.cmd]: Yep.  My mistake.

**Comment by TyOverby** [comment]
@jaredpar My CI build failed saying that 

```
09:32:45 ERROR: (netci.groovy, line 168) No signature of method: static jobs.generation.Utilities.GetFullJobName() is applicable for argument types: (java.lang.String, java.lang.String, java.lang.Boolean) values: [dotnet/roslyn, perf-correctness, false]
09:32:45 Possible solutions: getFullJobName(java.lang.String, java.lang.String, boolean), getFullJobName(java.lang.String, java.lang.String, boolean, java.lang.String)
```

I just copied your determinism code in the netci.groovy script, so I'm fairly sure that I things didn't mess things up *that* badly.

**Comment by TyOverby** [comment]
@dotnet-bot test ci please

File: netci.groovy [hunk 10]
```diff
@@ -151,3 +151,20 @@ set TMP=%TEMP%
...
+  def jobName = Utilities.getFullJobName(projectName, "perf_correctness", isPr)
+  def myJob = job(jobName) {
+    description('perf test correctness')
+    label('windows-roslyn')
+    steps {
+      batchFile(""".\\cibuild.cmd /testPerfCorrectness""")
+    }
+  }
+
+  def triggerPhraseOnly = true
```
Comment (jaredpar) [inline netci.groovy]: This should eventually be `false`.  When `true` it will only run when developers specifically ask for it in PR.

Comment (jaredpar) [inline netci.groovy]: To be clear: i think we should start with this as is.  We can work out the kinks in the job and eventually set it to `false`.

**Comment by TyOverby** [comment]
@jaredpar Huh, looks like I got it working.

**Comment by TyOverby** [comment]
can I have it run the new test job that I created?

**Comment by jaredpar** [comment]
Not yet.

**Comment by jaredpar** [comment]
Hmm, maybe @mmitche knows a way to run the generated job.  I don't htink there is a way yet.

**Comment by mmitche** [comment]
@TyOverby You can manually launch the created job by enabling it (the one under GenPRTest).  Make sure it's disabled after it runs though

**Comment by TyOverby** [comment]
@mmitche I'm guessing you mean this one? http://dotnet-ci.cloudapp.net/job/Private/job/dotnet_roslyn-internal/job/master/job/GenPRTest/job/dotnet_roslyn-internal_generator_prtest/

How do I get it to do a run?

**Comment by mmitche** [comment]
Ahh, first you do a 'test ci please'.  The disabled jobs are deleted every night so your run from yesterday is stale.

Then find the generated job you want, enable it and then "Build with Parameters"

**Comment by TyOverby** [comment]
@mmitche The last time I ran `test ci please` was only three hours ago though.

**Comment by mmitche** [comment]
Hmm well they got deleted:http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/GenPRTest/

Now, you may enable one of those builds and run it manually.

**Comment by mmitche** [comment]
also, your last comment retriggered it again :D

**Comment by TyOverby** [comment]
HAHAHA

Perfect

**Comment by mmitche** [comment]
You can still queue it while the other job is building

**Comment by TyOverby** [comment]
Ok, I enabled the project, started a build targeting this branch, and then disabled the project again.

**Comment by TyOverby** [comment]
The job run is here: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/GenPRTest/job/perf_correctness/

**Comment by TyOverby** [comment]
Dang, it looks like I can't get it to test a branch that is on my own fork

**Comment by mmitche** [comment]
Queue this one http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/GenPRTest/job/perf_correctness_prtest/

**Comment by mmitche** [comment]
It has instructions for running against your own fork

**Comment by TyOverby** [comment]
@mmitche Oh neat, I was looking at the non-pr based one.

**Comment by TyOverby** [comment]
[Yay; It succeeded!](http://dotnet-ci.cloudapp.net/job/dotnet_roslyn/job/master/job/GenPRTest/job/perf_correctness_prtest/1/console)  `test perf-correctness` will be a thing!

**Comment by mmitche** [comment]
NICE!  MERGE MERGE! :100:

File: netci.groovy (hunk 10 above)
Comment (TyOverby) [inline netci.groovy]: Agreed

**Comment by TyOverby** [comment]
@dotnet/roslyn-perf Hey, we have a new alias!  Also, please review; if everything looks good, I'll start writing documentation.

**Comment by jaredpar** [comment]
Jenkins changes LGTM

**Comment by jaredpar** [comment]
@KevinH-MS, @basoundr please review soon.  Need to get this job online to begin working with the perf team.

**Comment by basoundr** [comment]
LGTM. Please do check that all our VS tests in the closed repo run without any issues before checking-in. This is important

**Comment by jaredpar** [comment]
@basoundr what tests do you think could be affected by this: VSI, ETA, etc ...?  Given that ETA has passed I don't see how VSI could be at fault here.  This PR is already days behind due to lack of progress.  Want to get moving.

**Comment by TyOverby** [comment]
@jaredpar We have some pref tests that live in the closed repo.  I think there's one TAO test that goes through our runner.

**Comment by basoundr** [comment]
@Jaredpar The tests that we have in the open are command line tests. Since TAO is internal, we placed the VS based tests in the closed Roslyn. When we build Roslyn, tests from both open and closed gets copied into the tests directory for the perf tool to pick up. If this change breaks the running of the closed tests, then tool is broken we wont be getting our results.

**Comment by TyOverby** [comment]
@basoundr Good call on running the private tests.  All the tao-based tests crash immediately.

```
Unhandled Exception: Roslyn.Hosting.Test.Execution.Logging.PerfLogger+PerfLoggerException: Perf test encountered a Roslyn.Hosting.Test.Execution.Logging.PerfLogger+PerfLoggerException. See InnerException for more details. ---> Roslyn.Hosting.Test.Execution.Logging.PerfLogger+PerfLoggerException: Perf test encountered a System.IO.FileLoadException. See InnerException for more details. ---> System.IO.FileLoadException: Could not load file or assembly 'Microsoft.CodeAnalysis.Features, Version=42.42.42.42, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
Server stack trace:
   at Roslyn.VSHosting.Test.RoslynInProcessAPI.get_WaitingServiceAvailable()
   at Roslyn.VSHosting.Test.RoslynInProcessAPI.EnableTrackActiveTokens() in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\RoslynTaoActions\Actions\VisualStudio\Utilities\RoslynInProcessAPI.cs:line 114
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)
Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at Roslyn.VSHosting.Test.RoslynInProcessAPI.EnableTrackActiveTokens()
   at Roslyn.VSHosting.Test.RoslynActions.StartTargetAction.OnAfterStartupComplete() in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\RoslynTaoActions\Actions\VisualStudio\StartTargetAction.cs:line 16
   at Roslyn.VSHosting.Test.StartTargetAction.Execute(TaoEngine engine) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Actions\VisualStudio\StartTargetAction.cs:line 128
   at Tao.Engine.DebuggableActionExecutor.Execute() in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 139
   at Tao.Engine.DefaultActionExecutor.Execute() in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 36
   --- End of inner exception stack trace ---
   at Roslyn.Hosting.Test.Execution.Logging.PerfLogger.Tao.Engine.Logging.ITaoLogger.LogFailure(Exception ex) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\Logging\PerfLogger.cs:line 34
   at Tao.Engine.Logging.Log.Failure(Exception error) in f:\dd\src\Omni\Tao\Tao.Engine\Logging\Log.cs:line 95
   at Tao.Engine.DefaultActionExecutor.Execute() in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 40
   at Tao.Engine.DebuggableActionExecutor.ExecuteAction(TaoEngine engine, ActionToken actionToken) in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 106
   at Tao.Utilities.DecoratorExecutor.ExecuteTopLevelAction(TaoEngine engine, ActionToken actionToken) in f:\dd\src\Omni\Tao\Tao.Utilities\DecoratorExecutor.cs:line 50
   at Tao.Utilities.EtwExecutor.ExecuteTopLevelAction(TaoEngine engine, ActionToken actionToken) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\EtwExecutor.cs:line 16
   at Tao.Utilities.DecoratorExecutor.ExecuteAction(TaoEngine engine, ActionToken actionToken) in f:\dd\src\Omni\Tao\Tao.Utilities\DecoratorExecutor.cs:line 33
   at Tao.Engine.TaoEngine.ExecuteAction(ActionToken actionToken) in f:\dd\src\Omni\Tao\Tao.Engine\TaoEngine.cs:line 144
   at Tao.Engine.Execution.TestExecutionProviderDecorator.ExecuteAction(ActionToken action) in f:\dd\src\Omni\Tao\Tao.Engine\Execution\TestExecutor.cs:line 50
   at Tao.Engine.Execution.ActionBlock.Execute(ITestExecutionProvider provider) in f:\dd\src\Omni\Tao\Tao.Engine\Execution\ExecutionPlan.cs:line 362
   at Tao.Environment.Debugger.PausableExecutor.ExecuteTest(TaoTest test, ITestExecutionProvider executionVisitor, ITestBlockFactory blockFactory) in f:\dd\src\Omni\Tao\Tao.Environment\Debugger\PausableExecutor.cs:line 104
   at Tao.Engine.TaoEngine.ExecuteTest(TaoTest test) in f:\dd\src\Omni\Tao\Tao.Engine\TaoEngine.cs:line 60
   at Tao.Environment.CommandLineExecutor.Execute(String[] args, ArgumentHandler commandLineHandler, ArgumentHandler environmentHandler) in f:\dd\src\Omni\Tao\Tao.Environment\CommandLineExecutor.cs:line 146
   at Tao.Environment.CommandLineExecutor.Execute(String[] args, ArgumentHandler commandLineHandler) in f:\dd\src\Omni\Tao\Tao.Environment\CommandLineExecutor.cs:line 106
   at Roslyn.Hosting.Test.TaoEntryPoint.RunIterations(Arguments arguments, IEnumerable`1 files, CommonTaoEngine engine, CommandLineExecutor executor, String rootSuffix) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 659
   at Roslyn.Hosting.Test.TaoEntryPoint.RunPerfTestForSuffix(Arguments arguments, IEnumerable`1 files, CommonTaoEngine engine, CommandLineExecutor executor, String rootSuffix, IEnumerable`1& steps, String& scenarioGroupName) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 557
   at Roslyn.Hosting.Test.TaoEntryPoint.RunTestsForPerf(Arguments arguments, IEnumerable`1 files, CommonTaoEngine engine, CommandLineExecutor executor, Boolean profiling) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 522
   at Roslyn.Hosting.Test.TaoEntryPoint.RunTests(Arguments arguments, IEnumerable`1 files, Report report, Action`1 testResultCallback) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 440
   at Roslyn.Hosting.Test.TaoEntryPoint.MainWorker(String[] args) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 192
   --- End of inner exception stack trace ---
   at Roslyn.Hosting.Test.Execution.Logging.PerfLogger.Tao.Engine.Logging.ITaoLogger.LogFailure(Exception ex) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\Logging\PerfLogger.cs:line 34
   at Tao.Engine.Logging.Log.Failure(Exception error) in f:\dd\src\Omni\Tao\Tao.Engine\Logging\Log.cs:line 95
   at Roslyn.Hosting.Test.TaoEntryPoint.MainWorker(String[] args) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 216
   at Roslyn.Hosting.Test.TaoEntryPoint.EntryPoint() in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 62
   at Roslyn.Hosting.Test.TaoEntryPoint.RunSingleInstance() in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 55
   at Roslyn.Hosting.Test.TaoEntryPoint.Main(String[] args) in C:\Users\tyoverby\workspace\roslyn-internal\Closed\Hosting\Test\Execution\TaoEntryPoint.cs:line 38
```

**Comment by TyOverby** [comment]
@dotnet-bot test this please

**Comment by TyOverby** [comment]
This patch got delayed by the `check-branch.ps1` being broken and insta-failed my builds.

**Pull request merged**